Data/HashMap/Base.hs

```diff
@@ -1651,28 +1652,18 @@ filterMapAux onLeaf = go
     go (Full ary) = filterA ary fullNodeMask
     go (Collision h l1 l2 hm) = case (onLeaf (Leaf h l1), onLeaf (Leaf h l2)) of
       (Just (Leaf _ l1'), Just (Leaf _ l2')) -> Collision h l1' l2' $ go hm
-      (Just (Leaf _ l1'), Nothing) -> case go' hm of
-        Nothing -> Leaf h l1'
-        Just (Leaf _ l3', hm') -> Collision h l1' l3' hm'
-        _ -> error "Should not happen, can be fixed with refactoring I think."
-      (Nothing, Just (Leaf _ l2')) -> case go' hm of
-        Nothing -> Leaf h l2'
-        Just (Leaf _ l3', hm') -> Collision h l2' l3' hm'
-        _ -> error "Should not happen, can be fixed with refactoring I think."
-      (Nothing, Nothing) -> rehash h $ go hm
+      (Just (Leaf _ l1'), Nothing) -> go1 l1'
+      (Nothing, Just (Leaf _ l2')) -> go1 l2'
+      (Nothing, Nothing) -> case unconsHM (go hm) of
+        UnconsEmptyHM -> Empty
+        UnconsedHM l1' hm' -> case unconsHM hm' of
+          UnconsEmptyHM -> Leaf h l1'
+          UnconsedHM l2' hm'' -> Collision h l1' l2' hm''
       _ -> error "Should not happen, can be fixed with refactoring I think."
```

Comment:
I think you're probably right about refactoring, but that should *certainly* be in a different PR. In particular, I think we should be able to use `ptrEq` in `mapMaybeWithKey` to determine whether a value or subtree was changed. If not, we can reuse the old one. That *should* take care of `filterWithKey` and also win us some sharing if someone only changes some of the values in a map.